libkbfs/tlf_journal.go

```diff
@@ -1238,11 +1360,90 @@ func (j *tlfJournal) putMD(ctx context.Context, rmd *RootMetadata) (
 		return MdID{}, err
 	}
 
+	// Merge the unflushed paths.
+	if newUnflushedPaths != nil {
+		if _, ok := j.unflushedPaths[rmd.Revision()]; ok {
+			return MdID{}, fmt.Errorf("Revision %d already has unflushed paths",
+				rmd.Revision())
+		}
+		j.unflushedPaths[rmd.Revision()] = newUnflushedPaths
+	}
+
 	j.signalWork()
 
 	return mdID, nil
 }
 
+// prepUnflushedPaths returns a set of paths that were updated in this
+// revision.  The caller must NOT already hold `journalLock`.
+func (j *tlfJournal) prepUnflushedPaths(ctx context.Context,
+	rmd *RootMetadata) (map[string]bool, error) {
+	cpp := func() chainsPathPopulator {
+		j.journalLock.RLock()
+		defer j.journalLock.RUnlock()
+		return j.chainsPopulator
+	}()
+
+	// The unflushed paths haven't been initialized yet.
+	if cpp == nil {
+		return nil, nil
+	}
+
+	newUnflushedPaths := make(map[MetadataRevision]map[string]bool)
+	irmds := []ImmutableRootMetadata{
+		// Revision and timestamp don't matter for finding the
+		// unflushed paths.
+		MakeImmutableRootMetadata(rmd, fakeMdID(1), time.Now()),
+	}
+
+	err := j.addUnflushedPaths(ctx, irmds, cpp, newUnflushedPaths)
+	if err != nil {
+		return nil, err
+	}
+	if len(newUnflushedPaths) > 1 {
+		return nil, fmt.Errorf("%d unflushed revisions on a single put",
+			len(newUnflushedPaths))
+	}
+
+	return newUnflushedPaths[rmd.Revision()], nil
+}
+
+func (j *tlfJournal) putMD(ctx context.Context, rmd *RootMetadata) (
+	MdID, error) {
+	// Before we take the lock, get the unflushed paths ready (which
+	// might require taking the lock).
+	newUnflushedPaths, err := j.prepUnflushedPaths(ctx, rmd)
+	if err != nil {
+		return MdID{}, err
+	}
+	mdID, err := j.doPutMD(ctx, rmd, newUnflushedPaths)
+
+	// If someone initialized j.unflushedPaths, then try prepping the
+	// unflushed paths one more time.
+	if err == errRetryUnflushedPaths {
+		err := j.maybeWaitForUnflushedReady(ctx)
```

Comment:
Okay, so I think I understand how this all works. But it's pretty complicated, and I think it can be simpler. Also, I think we can avoid blocking the `putMD` path on initializing the unflushed paths, and the retry logic.

So I think we need a separate object, say `unflushedPathCache` that holds:

- a lock, protecting all variables.
- a state enum, which is either uninitialized, initializing, or initialized.
- a channel (closed when the state goes to initialized)
- unflushedPaths and chainsPopulator, non-nil only when the state is initialized.
- a queue of MD objects / revisions, non-empty only when the state is initializing.

Let `unflushedPathCache` have a `getUnflushedPaths`, which will (under [r-]lock) immediately return `unflushedPaths` if initialized, but return an error otherwise.

Also have a `startInitializeOrWait` method, which will:

1) take the lock, defer unlock
2) If the state is uninitialized, transition to initializing, initialize the channel, and return true. If the state is initialized, return false.
3) Otherwise, release the lock and block on the channel. When unblocked, grab the lock and do step 2 again (which should work this time).

Also have an `abortInitialization` method, which will transition from initializing to uninitialized and close the channel.

Have an `appendToCache` method, which takes an `ImmutableRootMetadata`. If the state is uninitialized, it does nothing. If it is initializing, it will append the MD to the queue. If it is initialized, it will process with the chainsPopulator and append to `unflushedPaths`. Have a similar `removeFromCache` method.

Have a `finishInitialization` method, which takes a set of unflushed paths. If state is initializing and the queue is empty, set the internal unflushed paths to the given one, close the channel, and return success. Otherwise, clear the queue and return it to the caller to process.

Then `getJournalStatusWithPaths` will do something like:

1) Call `getUnflushedPaths`. If it returns something, we're done. Otherwise, call `startInitializeOrWait`. Then call `startInitializeOrWait`. If false is returned, call `getUnflushedPaths`, which should succeed. Otherwise, defer `abortInitializiation` on error.
3) Initialize unflushed paths with the results from `getJournalStatusWithRange`.
4) Call `finishInitialization`. If it succeeds, we're done. Otherwise, if it returns more ops, process them, and repeat until done.

Then the `putMD` path will just have to call `appendToCache`, and likewise with the flush path.

What do you think?